src/stan_math_backend/Transform_Mir.ml

```diff
@@ -31,18 +62,37 @@ let rec switch_expr_to_opencl available_cl_vars (Expr.Fixed.({pattern; _}) as e)
     | _, UnsizedType.(UInt | UReal) -> e
     | _, _ -> to_matrix_cl e
   in
-  let check_type args (i, t) = Expr.Typed.type_of (List.nth_exn args i) = t in
-  let any_req_met args = List.exists ~f:(List.for_all ~f:(check_type args)) in
+  let check_type args (i, ad, t) =
+    Expr.Typed.type_of (List.nth_exn args i) = t
+    && UnsizedType.autodifftype_can_convert
+         (Expr.Typed.adlevel_of (List.nth_exn args i))
+         ad
+  in
+  let any_rest_met args =
```

Comment:
I would maybe just name this to `restricted` or `is_restricted`

Reply:
Agreed. Changed to is_restricted.